E2E: Add a --namespace hint which tests can choose to honor, and implement...

jayunit100 opened the pull request.

jayunit100 commented:
This is a solution to #6554  .

In follow on pull requests, we can

- update more tests to use the namespace hint, and base namespaces off of it
- consolidate the "randomeSuffix()" and "UUID" methods to all use something from `util.go` 

and other cleanups so that the ns logic is less ad hoc.  this is just a first step in that direction.

jayunit100 commented:
... Added one more commit to this, (unsquashed)... to update density tests to honor ns as well.  I think it should be a separate commit since its a separate change, but belongs in this PR.

satnam6502 commented:
Thank you kindly for this contribution. Personally, I think with an iron first we should insist on each e2e test creating its own namespace. I am wary of anything that distracts from that goal. @quinton-hoole

jayunit100 commented:
i see your point.  part of the intent here is to provide  a hint that can be used as prefix... but if we absolutely don't want users to be able to influence the name of this e2e namespaces at all,  i can close this guy.

satnam6502 commented:
I defer to the testing Ubermesiters @zmerlynn and @quinton-hoole

quinton-hoole commented:
Thanks jayunit100@.  Yes, we definitely need to clean up our act around namespaces for tests, and if you're willing to have a go at it, you have my full support.  I think that the correct approach is a variant of your suggestions, and goes roughly as follows:

1. Before every test invocation, the test infrastructure automatically selects/creates a well-defined list of unique namespaces, and passes them in to the test somehow (ginkgo BeforeEach() is one approach).  

2. All tests use only those namespaces.  e.g. if a given test only needs one namespace, it uses the first one.  If it needs more than one distinct namespace, it uses the first 'n' in the list etc.  tests/e2e/services.go provides an example of this sort of thing. No test should ever use any other namespace - that would constitute a bug (although those could be fixed in separate PR's).

3. After every test invocation, the test infrastructure automatically deletes everything out of all of the namespaces that it selected in step 1 above.  Again, ginkgo AfterEach() is one approach here.

The above ensures that:

1. Even if tests don't clean up properly after themselves, the framework will.
2. Each test invocation is guaranteed a clean namespace within which to operate, without any logic in the test.  This also enables us to run all of the tests in parallel (in theory, at least).
3. Tests become significantly simpler, because they don't each require relatively complex cleanup with necessary retry logic etc.  Most of the tests currently get this wrong, in many different places.

Any disagreement?  If not, go forth :-)

Q

jayunit100 commented:
that works for me !
Should we support the `--namespace`  ? or just statically create a very unique ns in the code like `org.googlecloud.e2e.testing[1-100]` ?

quinton-hoole commented:
No, I don't think that we should support --namespace, as that would enable/encourage people to pass non-unique namespaces into the tests.  So your latter suggestion applies.  There's code in tests/e2e/services.go that does a fairly reasonable job of this, although a simple piece of a GUID might be shorter and more unique.  You'll need to be careful of running up against length limits (e.g. external cloud load balancer names are currently constructed by concatenating the namespace and name of the service, and ELB names are limited to e.g. 64 chars in some cloud providers.  This is actually a bug in services that needs to be fixed.  There's an issue for it somewhere.

satnam6502 commented:
Whatever we do, please can we make sure it is relatively easy to run just a single test. If adding extra context in (e.g. a namespace supplied by the runner) makes it harder to carve out a test that can be run independently then that would be a shame. I am assuming this is not the case here -- but just in case -- I thought it worth mentioning.

jayunit100 commented:
gotcha, we will go with org.kubernetes....blah... as the default namespace for tests, and regex them with monotonically increasing integers like so

```
org.kubernetes.e2e0
org.kubernetes.e2e1
org.kubernetes.e2e2
org.kubernetes.e2e3
....

Sounds like a plan ?

timothysc commented:
If we can completely automate the creation and deletion and as much as possible that would be ideal.

derekwaynecarr commented:
This all sounds reasonable to me.

I think if we block on namespace deletion, e2e will appear a little slow until we move namespace_controller to use @lavalamp new controller framework.  Which I will look to do after his PR merges where the scheduler uses it as an example.

----- Original Message -----
From: "Timothy St. Clair" <notifications@github.com>
To: "GoogleCloudPlatform/kubernetes" <kubernetes@noreply.github.com>
Sent: Wednesday, April 8, 2015 4:22:53 PM
Subject: Re: [kubernetes] E2E: Add a --namespace hint which tests can choose to honor, and implement... (#6553)

If we can completely automate the creation and deletion and as much as possible that would be ideal. 
 

---
Reply to this email directly or view it on GitHub:
https://github.com/GoogleCloudPlatform/kubernetes/pull/6553#issuecomment-91025765

quinton-hoole commented:
satnam6502@ - yes, you will still be able to run individual tests exactly as today.  The Ginkgo framework automatically interjects and sets things up before each test.

jayunit100@ - remember that namespaces are local to a kubernetes cluster, so the org.kubernetes you propose is superfluous. I would prefer a simple UUID or timestamp.  Also, the monotonically increasing integers you propose are going to be challenging to do, as they require you to store that state (the last integer used) somewhere between runs - and runs may be on different machines etc.  This is something that GUID's (and suitably fine-grained timestamps) solve much better.    Something like https://godoc.org/code.google.com/p/go-uuid/uuid

Q

quinton-hoole commented:
@derekwaynecarr - regarding namespace deletion, we could defer that right to the end of the test run, and delete all namespaces in parallel if that speeds things up a bit.

jayunit100 commented:
hi @quinton-hoole  ........... re UUIDs vs sequences.... 

UUID's are, indeed, a hammer for solving collisions in systems lacking any sort of consensus mechanism... but... isnt the whole point of etcd to ensure that namespace existence (and other things of course) are consistent across the cluster ?

quinton-hoole commented:
Yes, in theory you could use etcd to ensure namespace uniqueness, but that will be more complicated to implement than a simple UUID or timestamp.

jayunit100 commented:
sorry,didnt mean to suggest using etcd.... red herring.  i guess im just making sure i understand things correctly.  let me try again :) .....

- all tests run in order 
- all tests request 1,2,3,4,...N namespaces when starting
- all tests delete 1,2,3,4,...N namespaces when completing

So in this case, how could we ever possibly have a collision ?

quinton-hoole commented:
Aah - now I understand your question better.  Thanks for clarifying.

A few things to bear in mind:

1. We plan to run multiple invocations of the e2e tests, perhaps against the same cluster, and probably in parallel.
2. Some resource names in backend cloud providers are based upon the namespace name.  So if you run the tests multiple times, perhaps on different machines, you don't want namespace name collisions.  In your proposed scheme, the same namespace names would be generated across multiple runs.
3. Yes, test cleanup has traditionally been a problem (it's actually quite hard to do 100% reliably), and makes for very flaky test results, so we'd like to separate it from the actual success or failure of the tests.  If test cleanup fails for whatever reason, it will leave cruft behind, but not cause later tests to fail.  We can solve the cruft problem separately.

Q

quinton-hoole commented:
Assigning back to jayunit100@.  Feel free to assign it back to me for review when you're ready.  And thanks again for taking this on! :-)

quinton-hoole commented:
For some reason GitHub won't allow me to assign ti to you.  Feel free to self-assign.

jayunit100 commented:
okay cool. thanks!

jayunit100 commented:
@quinton-hoole ....  In the Above PR, I've implemented the fix for networking.go test lifecycle that is minimal.  Then we can  generalize and improve  it for other tests in this specific PR.....

jayunit100 commented:
Hi folks just to be clear **don't merge** we're redoing this PR and then ill update....

quinton-hoole commented:
Hi jayunit100@ . Any progress on generalizing what you did in #6847 to make it usable across all e2e tests?  I'm finding myself doing one-off patches all over the place, and would rather fix this for once and for all across all tests.  I'm happy to take this on if you're busy.  Just want to make sure that we're not duplicating effort.

timothysc commented:
@quinton-hoole jay is at apachecon this week so feedback may be ... spotty.

jayunit100 commented:
Hey !

- Indeed I will look directly into this now,  apachecon has now wound
down;
- and I'm spinning my kube dev environment back up today.

I'll start working on this and probably will have a patch by friday.

On Thu, Apr 16, 2015 at 9:51 AM, Timothy St. Clair <notifications@github.com
> wrote:

> @quinton-hoole <https://github.com/quinton-hoole> jay is at apachecon
> this week so feedback may be ... spotty.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/6553#issuecomment-93739145>
> .
>



-- 
jay vyas

jayunit100 committed:
Generalized namespace creation pattern for test utils

jayunit100 commented:
Hi again kubernetes .  ok..  I've redone this branch so that it implements monotonic ids + randomSuffixes for `networking.go`.  From here we can update all the tests easily .

quinton-hoole commented on test/e2e/util.go:
```diff
@@ -81,6 +81,24 @@ func waitForPodCondition(c *client.Client, ns, podName, desc string, condition p
+	ginkgoNsIndex++
+	namespaceObj := &api.Namespace{
+		ObjectMeta: api.ObjectMeta{
+			Name:      fmt.Sprintf("e2e-tests-%v-%v-%v", baseName, ginkgoNsIndex, strconv.FormatInt(time.Now().Unix(), 10)),
```
As discussed previously, please use a UID here, and not a timestamp or counter.

quinton-hoole replied:
Also, I don't think that this function actually has anything to do with Ginkgo, so please remove that from the name. createUniqueNamespace() should be fine.

jayunit100 replied:
Ok to have uuid with timestamp?  That way there is useful information with the ns. Recall name spaces are quite fragile and will be some time before we can really rely on them to life cycle flawlessly.. so more info is better

quinton-hoole commented on test/e2e/util.go:
```diff
@@ -81,6 +81,24 @@ func waitForPodCondition(c *client.Client, ns, podName, desc string, condition p
+// createNS should be used by every test, note that we append a common prefix to the provided test name.
+func createGinkgoNS(baseName string, c *client.Client) (*api.Namespace, error) {
+	ginkgoNsIndex++
+	namespaceObj := &api.Namespace{
```
nit: Does the 'Obj' suffix serve any purpose here?  I'd prefer plain 'namespace'.

jayunit100 replied:
Well... it differentiates it from a string ref errence... which is the natural way one thinks about a name space .  Still prefer name space or ok---?

quinton-hoole commented on test/e2e/util.go:
```diff
@@ -81,6 +81,24 @@ func waitForPodCondition(c *client.Client, ns, podName, desc string, condition p
+		},
+		Status: api.NamespaceStatus{},
+	}
+	_, err := c.Namespaces().Create(namespaceObj)
```
Do you really need to explicitly create the namespace?  I'm a bit vague on how all of this works internally.  I suspect that you can just decide on the namespace names here, along the lines of how test/e2e/services.go currently works.

jayunit100 replied:
Yes you do but there are ways to launch kube so n's is created dynamically.  Anyways for tests it's good to do this step by step don't you think?

jayunit100 replied:
To clarify : some clusters might configure the create namespace automatically feature differently.  Imo We shouldn't rely on that for e2es.. Sounds ok?

quinton-hoole commented on test/e2e/networking.go:
```diff
@@ -94,29 +96,18 @@ var _ = Describe("Networking", func() {
 
 	AfterEach(func() {
-		By("Destroying namespace for this suite")
+		By(fmt.Sprintf("Destroying namespace for this suite %v", namespaceObj.Name))
```
We should consider moving most of the namespace creation stuff to BeforeSuite() and the namespace deletion to AfterSuite(), rather than duplicate it in every test.

jayunit100 replied:
Unfortunately the e2r suite is monolithic . There is only one .. so all name spaces would get created at once which makes this not really possible .

jayunit100 replied:
That is... even if some tests were skipped the name space would be created... hence the before test impl.  Excuse typos on a plane :)

jayunit100 commented:
Thanks for the feedback. In transit will reissue a pr  after I hear back on them When I land.

jayunit100 commented:
@quinton-hoole implemented your suggestions ; 

- changed util method name, also changed namespaceObj to namespace
- removed monotonic / timespace and just used UUIDs for namespace.  

Regarding other comments,  

- preserverd explict ns creation, since the auto create ns option is not gaurannteed to be on in all clusters
- BeforeTest/AfterTest since that would lead to a bunch of namespaces created unnecessarily.  

looks okay now?

jayunit100 commented:
(rebased, assuming the necessary changes  are completed now...just ping me if others needed ...

quinton-hoole commented:
Hi Jay.  I see you've still not addressed the comment about using a UID rather than a timestamp.  I'll wait until that's done before reviewing further.

jayunit100 commented:
@quinton-hoole random suffix is the one which is slated to be refactored as UUID.  I was going to do in a subsequent patch (you can read the TODO on it) however, ill gladly make that change in this patch if you want ...

jayunit100 committed:
Implement pure UUID suffix Functionality into namespace, without implementing randomSuffix() (yet).

jayunit100 commented:
okay, @quinton-hoole all set for a second look.  i just implemented UUID in the simplest way possible inline in the ns creation...  didnt worry about refactoring randomSuffix()  as thats a separate issue .

jayunit100 commented:
bump :) @quinton-hoole

timothysc commented:
@satnam6502 , @roberthbailey , @zmerlynn ^ this is blocking some other changes we want to make.

satnam6502 commented:
As Quinton said, I am also confused by the explicit namespace creation and this seems to be a different pattern from the other testes (although I am biased here since I think I might have established that).

quinton-hoole commented:
LGTM. Given that the namespace creation is only in one place per test file, that's fine.  We can easily reconsider later if necessary.

jayunit100 commented:
@satnam6502 explicit namespace creation gaurantees that e2e's run on clusters where namespaces aren't automatically created for you (i.e. where `KUBE_ADMISSION_CONTROL="--admission_control=NamespaceAutoProvision... "` is not set up  )

satnam6502 commented:
Can't that always be set for a test e2e cluster?

quinton-hoole commented:
I think it's good to decouple cluster creation from running e2e tests, i.e. it's good to be able to run e2e tests against any old cluster, which as Jay alludes, may or may not be set up to auto-provision namespaces.

satnam6502 commented:
Agreed.
